File: 2-keys-keyboard/solution.py
from typing import Dict


class Solution:
    def largestFactor(self, n: int) -> int:
        res = 1
        for num in range(2, n):
            if n%num == 0:
                res = max(res, num)
        return res

    def dp(self, n, memo=None):
        if memo is None:
            memo = {1:0}
        if n not in memo:
            largest_factor = self.largestFactor(n)
            memo[n] = int(n/largest_factor) + self.dp(largest_factor)
        return memo[n]

    def minSteps(self, n: int) -> int:
        return self.dp(n)

File: 2-keys-keyboard/solution.cpp
#include <vector>

using namespace std;

class Solution {
private:
  int largestFactor(int n) {
    int res = 1;
    for(int i=2; i<n; i++) if(n%i == 0) res = i;
    return res;
  }

public:
  int minSteps(int n) {
    vector<int> dp(n+1, 0);
    for(int i=2; i<n+1; i++) {
      int lf = largestFactor(i); dp[i] = i/lf + dp[lf];
    }
    return dp[n];
  }
};
